rfc/006-backfill-blocks.md

```diff
@@ -0,0 +1,155 @@
+network.                                            
+
+## Proposal
+
+A node, seeking to state sync, will find an adequate snapshot and offer it to
+the application with the following format:
+
+```golang
+type snapshot struct {
+ 	Height   uint64
+ 	Format   uint32
+ 	Chunks   uint32
+ 	Hash     []byte
+ 	Metadata []byte
+
+ 	trustedAppHash []byte // populated by light client
+}
+```
+
+The application will then be expected to parse information in the meta data
+regarding the backfill height and then return it in the response (alongside the
+app's decision):
+
+```proto
+message ResponseOfferSnapshot {
+  Result result = 1;
+  int64 backfill_height = 2;
+}
+```
+
+If all chunks are accepted and state sync is successful then the backfill height
+would be returned from state sync to the node so that it can execute the
+backfill process (via the blockchain reactor) to retrieve the necessary blocks.
+A backfill height of 0 would mean that this step is ignored.
+
+```golang
+go func() {
+		state, commit, backfill_height, err := ssR.Sync(stateProvider, config.DiscoveryTime)
+		if err != nil {
+			ssR.Logger.Error("State sync failed", "err", err)
+			return
+		}
+		err = stateStore.Bootstrap(state)
+		if err != nil {
+			ssR.Logger.Error("Failed to bootstrap node with new state", "err", err)
+			return
+		}
+		err = blockStore.SaveSeenCommit(state.LastBlockHeight, commit)
+		if err != nil {
+			ssR.Logger.Error("Failed to store last seen commit", "err", err)
+			return
+		}
+
+    // proposed method for backfilling blocks
+    bcR.BackfillBlocks(backfill_height)
+
+		if fastSync {
+			err = bcR.SwitchToFastSync(state)
+			if err != nil {
+				ssR.Logger.Error("Failed to switch to fast sync", "err", err)
+				return
+			}
+		} else {
+			conR.SwitchToConsensus(state, true)
+		}
+}
+```
+
+`BackfillBlocks` would be implemented similarly to fast sync itself, by
+requesting blocks and validating them by matching the hash of the new header
+to the LastBlockID hash in the trusted header. This would use the same block
+pool and thus would be exclusive with the fast sync process (only one can run
+at a time).
+
+For completeness, the node should also retrieve the `ValidatorSet`s,
+`ConsensusParam`'s and `ABCIResponse`'s for each of the heights. In alignment
+with the separation between state and block, it may makes sense that the state
+sync reactor rather than the blockchain reactor serve, request and verify this
+data.
+
+This could be done in one of two ways: use the light client via the
+RPC connection or creating a new channel to send this data through.
+
+However, this separation would require multiple reads to retrieve the header at
+each height and would make it difficult to coordinate with the process running
+on the blockchain reactor. Hence I would lean towards extending the blockchain
+reactor with new channels to request this data structure, to verify it and to
+persist it to the state store.
```

Comment:
We should consider this in the context of a P2P light client transport as well. We may end up implementing the same functionality both for the blockchain reactor and for a light client reactor, which seems unfortunate.

Reply:
Agree. It seems like there are several paths we can go about this. I will try enumerate them instead here to try make it clearer as to what might be the best path to take.

Reply:
Hmm or perhaps this is implementation-specific content that should more likely reside in an ADR

Reply:
Any reactor changes would involve additional P2P channels and messages, and those are covered by the spec.

Reply:
Okay but does that necessarily mean we want to discuss the possible implementation paths in this RFC?

Reply:
I guess there's no other medium with which we can do it in except in a RFC

Reply:
If you want, you could always submit a spec RFC for the protocol changes and an ADR for the implementation of those changes. However, I'm not sure if we really need an ADR, since the implementation will often follow from the protocol changes.